distributed/actor.py

```diff
@@ -227,34 +276,73 @@ class ActorFuture:
     Actor
     """
 
-    def __init__(self, q, io_loop, result=None):
-        self.q = q
-        self.io_loop = io_loop
-        if result:
-            self._cached_result = result
-        self.status = "pending"
+    @abc.abstractmethod
+    def result(self, timeout: str | timedelta | float | None = None) -> _T:
+        ...
+
+    @abc.abstractmethod
+    def done(self) -> bool:
+        ...
+
+    def __repr__(self) -> Literal["<ActorFuture>"]:
+        return "<ActorFuture>"
+
 
-    def __await__(self):
+@dataclass(frozen=True, eq=False)
+class EagerActorFuture(BaseActorFuture[_T]):
+    """Future to an actor's method call when an actor calls another actor on the same worker"""
+
+    _result: _T
+
+    def __await__(self) -> Generator[object, None, _T]:
+        return self._result
+        yield
+
+    def result(self, timeout: object = None) -> _T:
+        return self._result
+
+    def done(self) -> Literal[True]:
+        return True
+
+
+@dataclass(frozen=True, eq=False)
+class _OK(Generic[_T]):
+    _v: _T
+
+    def unwrap(self) -> _T:
+        return self._v
+
+
+@dataclass(frozen=True, eq=False)
+class _Error:
+    _e: Exception
+
+    def unwrap(self) -> NoReturn:
+        raise self._e
+
+
+@dataclass(frozen=False, eq=False)
+class ActorFuture(BaseActorFuture[_T]):
+    def __init__(self, io_loop: IOLoop):
+        self._io_loop = io_loop
+        self._event = _LateLoopEvent()
+        self._out: _Error | _OK[_T] | None = None
+
+    def __await__(self) -> Generator[object, None, _T]:
         return self._result().__await__()
 
-    def done(self):
-        return self.status != "pending"
+    def done(self) -> bool:
+        return self._event.is_set()
 
-    async def _result(self, raiseit=True):
-        if not hasattr(self, "_cached_result"):
-            out = await self.q.get()
-            if out["status"] == "OK":
-                self.status = "finished"
-                self._cached_result = out["result"]
-            else:
-                self.status = "error"
-                self._cached_result = out["exception"]
-        if self.status == "error":
-            raise self._cached_result
-        return self._cached_result
+    async def _result(self) -> _T:
+        await self._event.wait()
+        out = self._out
+        assert out is not None
+        return out.unwrap()
```

Comment:
Can you elucidate a bit more the purpose of these wrapper classes, as opposed to the more direct inspection of the result that there was previously? It seems like they are related to trying to get a chain of custody for the generic `_T`, but I'm not sure it really buys much since the setting of the result [here](https://github.com/dask/distributed/pull/5709/files#diff-095a1af29784e71235a0a6eaf77169c7b1c9eff5cb5cf84d69323dedb880805eR218) isn't checked, so we ultimately have an `_OK(Unknown)`.

Reply:
Yeah ideally this would have been done with TypedDict but they're not supported https://github.com/python/mypy/issues/3863

```python
class _OK(TypedDict, Generic[_T]):
    status: Literal["OK"]
    result: _T

class _Error(TypedDict):
    status: Literal["error"]
    exception: Exception
...
    def _set_result(self, out: _Error | _OK[_T]): ...
```

> It seems like they are related to trying to get a chain of custody for the generic _T, but I'm not sure it really buys much since the setting of the result [here](https://github.com/dask/distributed/pull/5709/files#diff-095a1af29784e71235a0a6eaf77169c7b1c9eff5cb5cf84d69323dedb880805eR218) isn't checked, so we ultimately have an _OK(Unknown).

I needed to draw the line somewhere of what's typed and what's not in this PR, and chose to type all the methods and classes of BaseActorFuture.  `_Error | _OK[_T] | None` is also needed for the internal state `ActorFuture` so I think it's worth it for now